python/tank/util/metrics.py

```diff
@@ -0,0 +1,391 @@
+            self._workers.append(worker)
+
+        self._dispatching = True
+
+    def stop(self):
+        """Instructs all worker threads to stop processing metrics."""
+        for worker in self.workers:
+            worker.halt()
+
+        self._dispatching = False
+        self._workers = []
+
+    @property
+    def dispatching(self):
+        """True if started and dispatching metrics."""
+        return self._dispatching
+
+    @property
+    def workers(self):
+        """A list of workers threads dispatching metrics from the queue."""
+        return self._workers
+
+    def _metrics_supported(self, sg_connection):
+        """Returns True if server supports the metrics api endpoint."""
+
+        # TODO: update the version number once the endpoint is available
+        self._metrics_ok = (
+            sg_connection.server_caps.version and
+            sg_connection.server_caps.version >= (7, 0, 0)
+        )
+
+        # TODO: remove after testing
+        self._metrics_ok = True
+        return self._metrics_ok
+
+
+class _MetricsDispatchWorkerThread(Thread):
+    """Worker thread for dispatching metrics to sg logging endpoint.
+
+    Once started this worker will dispatch logged metrics to the shotgun api
+    endpoint. The worker retrieves any pending metrics after the
+    `DISPATCH_INTERVAL` and sends them all in a single request to sg.
+
+    """
+
+    API_ENDPOINT = "api3/log_metrics/"
+    """SG API endpoint for logging metrics."""
+
+    DISPATCH_INTERVAL = 5
+    """Worker will wait this long between metrics dispatch attempts."""
+
+    DISPATCH_BATCH_SIZE = 10
+    """Worker will dispatch this many metrics at a time, or all if <= 0."""
+
+    def __init__(self, tk, engine):
+        """Initialize the worker thread.
+
+        :params tk: Toolkit api instance.
+        :params engine: Engine instance.
+
+        """
+
+        super(_MetricsDispatchWorkerThread, self).__init__()
+
+        self._tk = tk
+        self._engine = engine
+
+        # Make this thread a daemon. This means the process won't wait for this
+        # thread to complete before exiting. In most cases, proper engine
+        # shutdown should halt the worker correctly. In cases where an engine
+        # is improperly shut down, this will prevent the process from hanging.
+        self.daemon = True
+
+        # makes possible to halt the thread
+        self._halt_event = Event()
+
+    def run(self):
+        """Runs a loop to dispatch metrics that have been logged."""
+
+        # run until halted
+        while not self._halt_event.isSet():
+
+            # get the next available metric and dispatch it
+            try:
+                metrics = MetricsQueueSingleton().get_metrics(
+                    self.DISPATCH_BATCH_SIZE)
+                if metrics:
+                    self._dispatch(metrics)
+            except Exception, e:
+                self._engine.log_error("Error dispatching metrics: %s" % (e,))
```

Comment:
we have had issues with threading and logging in the past. If possible, I would personally keep the worker mute and dispatch any erroneous state back to the main thread if necessary. Some engines do thread safe logging, most don't. ALthough I may be overly cautious here :).